### PR TITLE
Fix shadow copy up to date check on startup

### LINKED ISSUE #48233: ANCM crashes with shadow copy enabled and the new deployment deleted a directory
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

I am using the shadow copy feature in .NET 7.  We got the following crash in the ANCM after deploying a change that deleted a directory under our `wwwroot` folder and the entire site was down with a 500 error until I added the empty directory back.

```
Application '/LM/W3SVC/6/ROOT' with physical root 'C:\Site\' failed to load coreclr. Exception message:
Unexpected exception: directory_iterator::directory_iterator: The system cannot find the path specified.: "C:\Site\wwwroot\lib"
```

However, I was able to eliminate the issue for subsequent deployments that might delete directories all together by using a pretty undocumented `cleanShadowCopyDirectory` setting found in #28357.

So it seems as if the ANCM is iterating on the shadow copied directory somewhere, but not checking if the folder actually exists in the newly deployed application.

I am not a C++ developer nor have I had a chance to actually run this code from source, but I have a hunch it is something to do with [`Environment::CheckUpToDate`](https://github.com/dotnet/aspnetcore/blob/main/src/Servers/IIS/AspNetCoreModuleV2/CommonLib/Environment.cpp#L206), which appears to be iterating directories without checking for existence.  If I am reading it correctly:
- It starts at the root of each directory (`source` = deployed app, `destination` = shadow copied)
- It iterates files/directories in the source.  If the item is a directory, recursively call `CheckUpToDate`, BUT with the `destination` directory as the `source` now - and it keeps flipping back and forth without checking whether the directory exists.  My guess is that the flow goes like this for each recursive call:
  - source: `<web root>\`, destination = `<shadow copy root>\` 
  - source: `<shadow copy root>\wwwroot`, destination = `<web root>\wwwroot`
  - source: `<web root>\wwwroot\lib`, destination = `shadow copy root>\wwwroot\lib`
    - `source` was set that way because it existed in `<shadow copy root>\wwwroot\lib`, but it does not actually exist in the newly deployed app.  `CheckUpToDate` attempts to iterate the directory and we crash. 

### Expected Behavior

If a directory is deleted in the deployed app, shadow copy should handle that and not crash.

### Steps To Reproduce

- Create an ASP.NET Core 7 application
- Set it up for shadow copy (but don't use `cleanShadowCopyDirectory`)
- Create a file under `wwwroot/lib` for example
- Deploy it to an IIS site (make sure the IIS site is set up for shadow copy as well with all the permissions)
  - We use `robocopy` to `publish` the site from CI and copy it into the site's root folder. 
- Run the site, see that it runs just fine
- Make some changes to your project:
  - Delete the `lib` directory, 
  - Make a change to C# code, re-compile (so as to simulate a new deployment)
- `publish` and re-deploy.
- See that it crashes when trying to access the site.  You will not be able to recover from this crash.

I can try to put together a sample repo if needed, but it seems to be as simple as that.  

### Exceptions (if any)

`directory_iterator::directory_iterator: The system cannot find the path specified.: <path to deleted directory>`

### .NET Version

2.2.200 (but this is on a server that only needs the runtime)

### Anything else?

(Note: this is our IIS Server and we don't keep the SDK up to date, we just install the runtime.  I don't feel like the SDK really has an impact here anyway - but let me know if I am misunderstanding that.)

NET Core SDK (reflecting any global.json):
 Version:   2.2.300
 Commit:    73efd5bd87

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.17763
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\2.2.300\

Host:
  Version:      7.0.2
  Architecture: x64
  Commit:       d037e070eb

.NET SDKs installed:
  2.2.300 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.All 2.2.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  [Microsoft.AspNetCore.App](http://microsoft.aspnetcore.app/) 2.2.5 [C:\Program Files\dotnet\shared\[Microsoft.AspNetCore.App](http://microsoft.aspnetcore.app/)]
  [Microsoft.AspNetCore.App](http://microsoft.aspnetcore.app/) 3.1.14 [C:\Program Files\dotnet\shared\[Microsoft.AspNetCore.App](http://microsoft.aspnetcore.app/)]
  [Microsoft.AspNetCore.App](http://microsoft.aspnetcore.app/) 5.0.5 [C:\Program Files\dotnet\shared\[Microsoft.AspNetCore.App](http://microsoft.aspnetcore.app/)]
  [Microsoft.AspNetCore.App](http://microsoft.aspnetcore.app/) 6.0.9 [C:\Program Files\dotnet\shared\[Microsoft.AspNetCore.App](http://microsoft.aspnetcore.app/)]
  [Microsoft.AspNetCore.App](http://microsoft.aspnetcore.app/) 7.0.2 [C:\Program Files\dotnet\shared\[Microsoft.AspNetCore.App](http://microsoft.aspnetcore.app/)]
  [Microsoft.NETCore.App](http://microsoft.netcore.app/) 2.2.5 [C:\Program Files\dotnet\shared\[Microsoft.NETCore.App](http://microsoft.netcore.app/)]
  [Microsoft.NETCore.App](http://microsoft.netcore.app/) 3.1.14 [C:\Program Files\dotnet\shared\[Microsoft.NETCore.App](http://microsoft.netcore.app/)]
  [Microsoft.NETCore.App](http://microsoft.netcore.app/) 5.0.5 [C:\Program Files\dotnet\shared\[Microsoft.NETCore.App](http://microsoft.netcore.app/)]
  [Microsoft.NETCore.App](http://microsoft.netcore.app/) 6.0.9 [C:\Program Files\dotnet\shared\[Microsoft.NETCore.App](http://microsoft.netcore.app/)]
  [Microsoft.NETCore.App](http://microsoft.netcore.app/) 7.0.2 [C:\Program Files\dotnet\shared\[Microsoft.NETCore.App](http://microsoft.netcore.app/)]

Other architectures found:
  x86   [C:\Program Files (x86)\dotnet]
    registered at [HKLM\SOFTWARE\dotnet\Setup\InstalledVersions\x86\InstallLocation]

Environment variables:
  Not set

global.json file:
  Not found

Learn more:
  https://aka.ms/dotnet/info

Download .NET:
  https://aka.ms/dotnet/download

### PR DESCRIPTION
**Pull request by BrennanConroy**
Fixes https://github.com/dotnet/aspnetcore/issues/48233

If using shadow copy in ANCM and you shutdown the app (e.g. app_offline.htm) then delete a subfolder (with a dll), starting up the app can cause a crash that can only be resolved by adding back the deleted folder. Or resolved by deleting the shadow copy folder (either manually or with `cleanShadowCopyDirectory` setting turned on).

The issue submitter helpfully pointed out the code that was problematic, we basically swapped which directory is the "main" directory every recursion of the up to date function. The fix is simple, keep the app directory as the main directory.

Test added that repros the scenario.

### CODE REVIEW

File: src/Servers/IIS/AspNetCoreModuleV2/CommonLib/Environment.cpp [hunk 1]
```diff
@@ -231,7 +231,7 @@ bool Environment::CheckUpToDate(const std::wstring& source, const std::filesyste
             auto sourceInnerDirectory = std::filesystem::directory_entry(path);
             if (sourceInnerDirectory.path() != directoryToIgnore)
             {
-                CheckUpToDate(destination / path.path().filename(), path.path(), extension, directoryToIgnore);
+                CheckUpToDate(path.path(), destination / path.path().filename(), extension, directoryToIgnore);
```
Comment (amcasey) [inline src/Servers/IIS/AspNetCoreModuleV2/CommonLib/Environment.cpp:234]: If this were C#, I would definitely use named arguments.  Comments might still be worthwhile (i.e. `/*source*/` and `/*destination*/`).

**Review by amcasey — APPROVED** [review]
Kudos, @someguy20336!

File: src/Servers/IIS/AspNetCoreModuleV2/CommonLib/Environment.cpp (hunk 1 above)
Comment (BrennanConroy) [reply src/Servers/IIS/AspNetCoreModuleV2/CommonLib/Environment.cpp:234]:
```suggestion
                CheckUpToDate(/*source*/ path.path(), /*destination*/ destination / path.path().filename(), extension, directoryToIgnore);
```

**Comment by antmjones** [comment]
Is there any possibility of this fix being included in a 8.0.x point release? I was caught out by this today and it was far from immediately obvious what had gone wrong (thanks again to @someguy20336 for the detailed info in the original report!)

I now know there's a workaround with setting ```cleanShadowCopyDirectory``` to true, but presumably this is going to have an impact on application startup time (given it will prevent skipping copying of files where the last write time is unchanged - https://github.com/dotnet/dotnet/blob/v8.0.4/src/aspnetcore/src/Servers/IIS/AspNetCoreModuleV2/CommonLib/Environment.cpp#L187).

**Pull request merged**